pyop2/codegen/representation.py

```diff
@@ -0,0 +1,561 @@
+import numbers
+import itertools
+from functools import partial
+from collections import defaultdict
+from pyop2.utils import cached_property
+import numpy
+from abc import ABCMeta
+
+
+from gem.node import Node as NodeBase
```

Comment:
Question, this makes tsfc (gem) a dependency of PyOP2. Do we want this? given that gem.node is so small, do we spin it out (or copy or something else?)

Reply:
Agreed and copied gem.node over.

Reply:
`gem.node` and `gem.utils` are standalone tools, they do no depend on the rest of GEM.